src/eve/pattern_matching.py

```diff
@@ -0,0 +1,159 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+import inspect
+from dataclasses import dataclass
+from functools import singledispatch
+from types import ModuleType
+from typing import Any, Iterator
+
+
+@dataclass(frozen=True)
+class ObjectPattern:
+    """Class to pattern match general objects.
+
+    A pattern matches an object if it is an instance of the specified
+    class and all attributes of the pattern (recursively) match the
+    objects attributes.
+
+    >>> class Foo:
+    ...    def __init__(self):
+    ...        self.bar = 1
+    ...        self.baz = 2
+    >>> assert ObjectPattern(Foo, {"bar": 1}).match(Foo())
+    """
+
+    cls: type
+    attrs: dict[str, Any]
+
+    def match(self, other: Any, raise_exception: bool = False) -> bool:
+        """Return if object pattern matches `other` using :func:`get_differences`.
```

Comment:
```suggestion
        """Return ``True`` if object pattern matches ``other`` using :func:`get_differences`.
```

Reply:
Mh, not a fan of this, but I've changed it. If I would be nit picking what you are proposing here is not as precise because it is unspecified what happens if the pattern does not match.

Reply:
It's a minor thing. You can change it back if you prefer, but please use proper restructured text markups.